### ISSUE TITLE
Use Configurator declared in ServerEndpointConfig over one declared in the @ServerEndpoint annotation

### ISSUE THREAD

**Issue by marctinkler** [eclipse/jetty.project#2559]
I believe there is a bug in org.eclipse.jetty.websocket.jsr356.server.AnnotatedServerEndpointConfig that is affecting our ability to create Guice injected websocket Configurators.  

Here is the problem, when registering a websocket via a call to ServerContainer.addEndpoint( ServerEndpointConfig ), you'd expect it to use *everything* configured in the ServerEndpointConfig (and not use the annotation, (even if the Endpoint you are adding is annotated with @ServerEndpoint).   And indeed, the AnnotatedServerEndpointConfig class makes sure to prioritize the "BaseConfig" over the annotation, in all but one case:

Here's a (correct) example starting on line 64:

```java
// Decoders (favor provided config over annotation)
        if (baseConfig != null && baseConfig.getDecoders() != null && baseConfig.getDecoders().size() > 0)
        {
            this.decoders = Collections.unmodifiableList(baseConfig.getDecoders());
        }
        else
        {
            this.decoders = Collections.unmodifiableList(Arrays.asList(anno.decoders()));
        }
```

However, when it comes to the Configurator, it prefers the *annotation* over the provided config:

starting on line 115:

```java
if (anno.configurator() == ServerEndpointConfig.Configurator.class)
        {
            if (configr != null)
            {
                cfgr = configr;
            }
            else
            {
                cfgr = new ContainerDefaultConfigurator();
            }
        }
        else
        {
            try
            {
                cfgr = anno.configurator().getDeclaredConstructor( ).newInstance();
            }
            catch (Exception e)
            {
                StringBuilder err = new StringBuilder();
                err.append("Unable to instantiate ClientEndpoint.configurator() of ");
                err.append(anno.configurator().getName());
                err.append(" defined as annotation in ");
                err.append(anno.getClass().getName());
                throw new DeploymentException(err.toString(),e);
            }
        }
```

I believe this is a bug, I think (based on the design of the class), this would be correct:

```java
       if (configr != null)
        {
                cfgr = configr;
        }
        else if (anno.configurator() == ServerEndpointConfig.Configurator.class)
        {
            cfgr = new ContainerDefaultConfigurator();
        }
        else
        {
            try
            {
                cfgr = anno.configurator().getDeclaredConstructor( ).newInstance();
            }
            catch (Exception e)
            {
                StringBuilder err = new StringBuilder();
                err.append("Unable to instantiate ClientEndpoint.configurator() of ");
                err.append(anno.configurator().getName());
                err.append(" defined as annotation in ");
                err.append(anno.getClass().getName());
                throw new DeploymentException(err.toString(),e);
            }
        }
```

**Comment by joakime** [comment]
Thanks for the report.  It does seem off.
That part of the code was tricky, as the javax.websocket spec doesn't make it clear which behavior to actually use in this scenario.

**Comment by joakime** [comment]
Added testcase, and fix in commit bc16b6e

**Comment by gregw** [comment]
This change broke some test: https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/jetty-9.4.x/339/tests/
